File: src/main/java/com/bss/inc/cameraon/licensing/impl/workers/ChecksumWorker.java
package com.bss.inc.cameraon.licensing.impl.workers;

import com.bss.inc.cameraon.licensing.interfaces.IChecksumWorker;
import com.bss.inc.cameraon.licensing.utils.ProductKeyUtils;

public class ChecksumWorker implements IChecksumWorker {@Override public String buildProductKeyChecksum(String productKey) {if ((productKey != null) && (productKey.trim().length() > 0)) {productKey = productKey.trim();int left = 0x0056;int right = 0x00AF;for (int n = 0; n < productKey.length(); n++ ) {right = right + (byte) productKey.charAt(n);if (right > 0x00FF) {right -= 0x00FF;}left += right;if (left > 0x00FF) {left -= 0x00FF;}}return ProductKeyUtils.buildHexStr(4, (left << 8) + right);}return null;}@Override public boolean verifyProductKeyChecksum(String productKey) {if ((productKey != null) && (productKey.trim().length() > 0)) {productKey = productKey.trim();if (productKey.length() > 4) {final String checkSum = productKey.substring(productKey.length() - 4);productKey = productKey.substring(0, productKey.length() - 4);return checkSum.equals(buildProductKeyChecksum(productKey));}}return false;}}

File: BSSLF/src/main/java/com/weilerhaus/productKeys/ChecksumWorker.cpp
#include <iostream>
#include <string>
#include <vector>
#include "ProductKeyUtils.h"
#include "ChecksumWorker.h"
#include "StringUtils.h"

std::string ChecksumWorker::buildProductKeyChecksum(std::string productKey) {
    if((productKey != nullptr) && (StringUtils::trim(productKey).length() > 0)) {
        productKey = StringUtils::trim(productKey);
        int left = 86;
        int right = 175;
        for(int n=0; n < productKey.length(); n++) {
            right = right + (int)productKey[n];
            if (right > 255) {right -= 255;}
            left += right;
            if (left > 255) {left -= 255;}
        }
        bhs(4, (left << 8) + right);
    }
    return nullptr;
}
bool ChecksumWorker::verifyProductKeyChecksum(std::string productKey) {
    if((productKey != nullptr) && (StringUtils::trim(productKey).length() > 0)) {
        productKey = productKey.trim();
        if (productKey.length() > 4)
        {
            std::string checkSum = productKey.substr(productKey.length() - 4);
            productKey = productKey.substr(0,productKey.length() - 4);
            return (checkSum.compare(buildProductKeyChecksum(productKey)) != 0);
        }
    }
    return false;
}
